### PR TITLE
Add support for match prefix-set

### PR DESCRIPTION
**Pull request by mjethanandani**
This is one in a series of commits to address the issues listed in issue #168.

### CODE REVIEW

File: src/yang/ietf-bgp-policy.yang [hunk 1]
```diff
@@ -21,7 +21,7 @@ module ietf-bgp-policy {
   organization
     "IETF IDR Working Group";
   contact
-    "WG Web:   <http://tools.ietf.org/wg/idr>
+    "WG Web:   <http://datatracker.ietf.org/wg/idr>
```
Comment (jhaas-pfrc) [inline src/yang/ietf-bgp-policy.yang:24]: Repeating comments from before, it's ideal to keep unrelated bits of diff from pull requests.  This probably should go under a separate issue to update the copyright dates.

File: src/yang/ietf-bgp-policy.yang [hunk 2]
```diff
@@ -61,31 +61,33 @@ module ietf-bgp-policy {
...
+   * typedef statements
+   */
 
   typedef bgp-set-community-option-type {
     type enumeration {
       enum add {
         description
           "Add the specified communities to the existing
-           community attribute";
+           community attribute.";
```
Comment (jhaas-pfrc) [inline src/yang/ietf-bgp-policy.yang:73]: Same comment for the unrelated punctuation cleanup

File: src/yang/ietf-bgp-policy.yang [hunk 3]
```diff
@@ -248,6 +253,92 @@ module ietf-bgp-policy {
...
+
+          leaf name {
+            type string;
+            description
+              "Name of the prefix set used to reference the set in
+               match conditions.";
+          }
+
+          leaf mode {
+            type enumeration {
```
Comment (jhaas-pfrc) [inline src/yang/ietf-bgp-policy.yang:275]: What does our extension model look like for updating this?
For example, while matching ipv4 and ipv6 prefixes is a core common application, what happens if you want your policy to match a vpnv4 prefix?  A rt-constrain prefix?  Etc.

File: src/yang/ietf-bgp-policy.yang [hunk 4]
```diff
@@ -248,6 +253,92 @@ module ietf-bgp-policy {
...
+                  "Prefix set contains mixed IPv4 and IPv6 prefixes";
+              }
+            }
+            description
+              "Indicates the mode of the prefix set, in terms of
+               which address families (IPv4, IPv6, or both) are
+               present. The mode provides a hint, but the device
+               must validate that all prefixes are of the indicated
+               type, and is expected to reject the configuration if
+               there is a discrepancy.  The 'mexed' mode may not be
```
Comment (jhaas-pfrc) [inline src/yang/ietf-bgp-policy.yang:295]: typo - mexed

File: src/yang/ietf-bgp-policy.yang [hunk 5]
```diff
@@ -248,6 +253,92 @@ module ietf-bgp-policy {
...
+                 prefix may be either IPv4 or IPv6, most
+                 implementations require all members of the prefix
+                 set to be the same address family. Mixing address
+                 types in the same prefix set is likely to cause an
+                 error.";
+            }
+
+            leaf masklength-range {
+              type string {
+                pattern '^(([0-9]+\.\.[0-9]+)|exact)$';
```
Comment (jhaas-pfrc) [inline src/yang/ietf-bgp-policy.yang:325]: What do people feel about "longer" and "or-longer" as additional easy forms?

Comment (jhaas-pfrc) [reply src/yang/ietf-bgp-policy.yang:325]: Also, regexes can't count.  General advice is that the first part of the range must be less than or equal to the second part of the range.  How do we want to note that?

**Review by jhaas-pfrc — CHANGES_REQUESTED** [review]

File: src/yang/ietf-bgp-policy.yang (hunk 3 above)
Comment (keyur-arrcus) [reply src/yang/ietf-bgp-policy.yang:275]: We will just have you augment it. :(

File: src/yang/ietf-bgp-policy.yang (hunk 5 above)
Comment (keyur-arrcus) [reply src/yang/ietf-bgp-policy.yang:325]: That is correct. my suspicion is that the range check would be hard. *If* we want, we can move it in the validation part.

File: src/yang/ietf-bgp-policy.yang (hunk 3 above)
Comment (jhaas-pfrc) [reply src/yang/ietf-bgp-policy.yang:275]: Right.  See prior discussions about the issues with augmenting for enumerations.  You can end up with inconsistent enumerations as a side effect.

Mahesh had a pattern that moved us to identities for the community types.  That might be appropriate here?

File: src/yang/ietf-bgp-policy.yang (hunk 5 above)
Comment (jhaas-pfrc) [reply src/yang/ietf-bgp-policy.yang]: To make this a concrete request:
add named options longer and or-longer
add to the description that number first number must be <= second number and that first number must be >= the prefix-length.

**Comment by mjethanandani** [comment]
@jhaas-pfrc, do the latest set of changes address your comments?

File: src/yang/ietf-bgp-policy.yang (hunk 1 above)
Comment (mjethanandani) [reply src/yang/ietf-bgp-policy.yang:24]: The tools are flagging the Copyright date as an issue, and they use the last modified date to decide whether the date should be updated.

**Review by keyur-arrcus — APPROVED** [review]
looks good so far.

File: src/yang/ietf-bgp-policy.yang (hunk 5 above)
Comment (mjethanandani) [reply src/yang/ietf-bgp-policy.yang]: > What do people feel about "longer" and "or-longer" as additional easy forms?

I added "longer" and "or-longer" as additional options in the mask length options, but examples on the web seem to indicate use of "le" or "ge". What is preferable?

**Comment by mjethanandani** [comment]
BTW, these changes have been validated with the example in PR #175.

File: src/yang/ietf-bgp-policy.yang [hunk 6]
```diff
@@ -272,6 +300,95 @@ module ietf-bgp-policy {
...
+          key "ip-prefix masklength-range";
+          description
+            "List of prefixes in the prefix set.";
+
+          leaf ip-prefix {
+            type inet:ip-prefix;
+            mandatory true;
+            description
+              "The prefix member in CIDR notation. While the
+                prefix may be either IPv4 or IPv6, most
```
Comment (jhaas-pfrc) [inline src/yang/ietf-bgp-policy.yang:352]: Indentation issue.

File: src/yang/ietf-bgp-policy.yang [hunk 7]
```diff
@@ -272,6 +300,95 @@ module ietf-bgp-policy {
...
+        }
+      }
+
+      container prefixes {
+        description
+          "Enclosing container for the list of prefixes in a policy
+           prefix list.";
+
+        list prefix {
+          key "ip-prefix masklength-range";
```
Comment (jhaas-pfrc) [inline src/yang/ietf-bgp-policy.yang:343]: A question for Keyur is do we think that implementations that support disjoint overlapping prefixes with range have consistent implementation for lookup.

**Review by jhaas-pfrc — APPROVED** [review]
